File: rubylet-ee/src/main/java/com/commongroundpublishing/rubylet/Servlet.java
package com.commongroundpublishing.rubylet;

import static com.commongroundpublishing.rubylet.Util.loadInstance;

import java.io.IOException;

import javax.servlet.ServletConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;

public class Servlet implements javax.servlet.Servlet {
    
    private javax.servlet.Servlet childServlet;

    public void init(ServletConfig config) throws ServletException {
        childServlet = loadInstance("com.commongroundpublishing.rubylet.jruby.RubyServlet",
                                    javax.servlet.Servlet.class);
        childServlet.init(config);
    }
    
    public void destroy() {
        getChild().destroy();
    }

    public javax.servlet.Servlet getChild() {
        return Util.assertNotNull(childServlet);
    }
    
    /**
     * Forward the http request to child.
     */
    public void service(ServletRequest req, ServletResponse resp)
        throws IOException, ServletException
    {
        getChild().service(req, resp);
    }
    
    public final String getServletInfo() {
        return getChild().getServletInfo();
    }

    public ServletConfig getServletConfig() {
        return getChild().getServletConfig();
    }

}

File: rubylet-ee/examples/demo_app/src/main/ruby/lib/demo/servlet.rb
module Demo
  class Servlet
    include Java::JavaxServlet::Servlet

    attr_reader :servlet_config
    
    # @param [javax.servlet.ServletConfig] servletConfig
    def init(servletConfig)
      @servlet_config = servletConfig
    end
    
    def destroy
      # no-op
    end

    alias :servlet_info :to_s

    # @param [javax.servlet.ServletRequest] req a Java servlet request,
    # assumed to be an HttpServletRequest
    #
    # @param [javax.servlet.ServletResponse] resp a Java servlet response
    # assumed to be an HttpServletResponse
    def service(req, resp)
      resp.setStatus(200)
      {
        'Content-Type' => 'text/plain'
      }.each { |k,v| resp.setHeader k, v }
      writer = resp.getWriter
      begin
        writer.println 'Hello, World, direct Ruby impl of javax.servlet.Servlet'
      ensure
        writer.close
      end
    end
  end
end
